include/seastar/websocket/server.hh

```diff
@@ -156,6 +156,16 @@ public:
  */
 class connection : public boost::intrusive::list_base_hook<> {
     using buff_t = temporary_buffer<char>;
+    /*!
+     * \brief Internal error that is thrown when there is no more data to read.
+     */
+    class eof_exception : public std::exception {
+    public:
+        eof_exception() {}
+        virtual const char* what() const noexcept {
+            return "Websocket connection: stream closed";
+        }
+    };
```

Comment:
EOF should not raise an exception, since they are exceedingly slow. EOF should be propagated in the normal control flow (perhaps as an empty temporary_buffer, or closing the queue).

Reply:
Technically, EOF does not **raise** an exception here, but dutifully creates an exception and passes it right to the exceptional future by calling `queue::abort()` (see line 174). The exception is only thrown because there's no other way of distinguishing why the queue got aborted, and I think that it's a valid use case - `queue::abort()` also the only way you can let `queue::pop_eventually()` know that it should resolve without waiting for more data.

@avikivity do you have any other solution to suggest? Ideally we should be able to inspect the reason for aborting without having to throw, and there's a whole mailing list discussion about introducing `seastar::immediate` for that purpose, but it's not there yet.

Reply:
Perhaps we could extend the interface of `queue` to have a way of soft-aborting with a given sentinel value instead of an exception. Then implementing eof would be trivial by calling `queue::soft_abort(empty_buffer_here)`. It sounds very specific to this particular use-case though, and it's hard to judge if such a sentinel value is generic enough to put it into queue's interface.

Reply:
hm, it sounds like `queue::push` is a great approximation of `soft_abort`.

Reply:
soft_abort() -> close() (and popping a disengaged optional). Probably need a new class for it.